syntax = "proto3";

package core;
option go_package = "./proto;core";

import "payment_task.proto";
import "payment_plan.proto";
import "accounts.proto";
import "transactions.proto";

service Core {
  rpc GetPaymentPlan(GetPaymentPlanRequest) returns (GetPaymentPlanResponse);
  //  CRUD METHODS
  // Account
  rpc CreateAccount(CreateAccountRequest) returns (Account);
  rpc GetAccount(GetAccountRequest) returns (Account);
  rpc ListAccounts(ListAccountRequest) returns (ListAccountResponse);
  rpc UpdateAccount(UpdateAccountRequest) returns (Account);
  rpc DeleteAccount(DeleteAccountRequest) returns (DeleteAccountResponse);
  // Transaction
  rpc CreateTransaction(CreateTransactionRequest) returns (Transaction);
  rpc GetTransaction(GetTransactionRequest) returns (Transaction);
  rpc ListTransactions(ListTransactionRequest) returns (ListTransactionResponse);
  rpc UpdateTransaction(UpdateTransactionRequest) returns (Transaction);
  rpc DeleteTransaction(DeleteTransactionRequest) returns (DeleteTransactionResponse);
  //Payment Task
  rpc CreatePaymentTask(CreatePaymentTaskRequest) returns (PaymentTask);
  rpc GetPaymentTask(GetPaymentTaskRequest) returns (PaymentTask);
  rpc ListPaymentTasks(ListPaymentTaskRequest) returns (ListPaymentTaskResponse);
  rpc UpdatePaymentTask(UpdatePaymentTaskRequest) returns (PaymentTask);
  rpc DeletePaymentTask(DeletePaymentTaskRequest) returns (DeletePaymentTaskResponse);
}

message GetPaymentPlanRequest {
  repeated string payment_tasks_ids = 1;
}

message GetPaymentPlanResponse {
  repeated PaymentPlan payment_plans = 1;
}
